### ISSUE TITLE
[Quantization]: Diffusion Models (Interface & Research RoadMap)

### ISSUE THREAD

**Issue by lishunyang12** [vllm-project/vllm-omni#896]
## 1. Motivation
As ```vllm-omni``` expands to support large-scale Diffusion Transformers (DiTs) like **Wan2.2 (14B)** and **Z-Image**, memory and compute efficiency become critical bottlenecks. While vLLM has mature quantization support for LLMs (AWQ, GPTQ, FP8), these are not yet exposed or optimized for the ```omni``` modality pipeline.

### This RFC proposes:

- A **unified interface** for users to apply quantization to diffusion models.

- A **research roadmap** to adapt existing kernels for DiT-specific architectures (e.g., 2D/3D positional embeddings, Modulation layers).

## 2. Design: User Interface
The design follows the "Configuration over Code" philosophy, allowing users to enable quantization via CLI or Python without modifying model code.

### 2.1 Python SDK Interface
We propose adding a ```quantization_config``` parameter to the ```Omni``` entrypoint. This aligns with ```vllm.LLM``` but adds specific controls for the diffusion pipeline (e.g., quantizing the DiT while keeping the VAE in float32).

``` python
from vllm_omni import Omni

# Example: Loading Wan2.1-14B with FP8 quantization
omni = Omni(
    model="Wan-AI/Wan2.1-T2V-14B",
    
    # NEW: Quantization Configuration
    quantization="fp8",           # Options: "awq", "gptq", "fp8", None
    kv_cache_dtype="fp8",         # Compress the attention cache
    
    # Fine-grained control (Future P1)
    # Allows keeping sensitive components in higher precision
    quantization_config={
        "ignore_patterns": ["vae", "text_encoder"], 
        "modules_to_not_convert": ["modulation_head"] 
    }
)
```

### 2.2 CLI Interface (**Serving**)
For the OpenAI-compatible server and offline scripts, we introduce standard flags.

Command:

``` Bash
python -m vllm_omni.entrypoints.openai.api_server \
    --model Wan-AI/Wan2.1-T2V-14B \
    --quantization awq \
    --kv-cache-dtype fp8_e5m2
```

### 2.3 Model Config (```config.json```)
The system will auto-detect quantization if the model artifact is pre-quantized.

``` JSON
// config.json inside the model folder
{
  "_name_or_path": "Wan2.1-14B-AWQ",
  "architectures": ["WanModel"],
  "quantization_config": {
    "quant_method": "awq",
    "bits": 4,
    "group_size": 128,
    "zero_point": true
  }
}
```
## 3. Future Research & Implementation Plan
This section outlines the technical roadmap for enabling the interface above.

### Zero-Shot Integration (P0)
_Goal: Re-use existing vLLM kernels for DiT Linear layers._

- **Mapping** ```nn.Linear```: The primary task is to map DiT layers (e.g., ```WanAttention```,```ZImageBlock```) to ```vllm.model_executor.layers.linear.RowParallelLinear```.
- **Weight Packing**: Create conversion scripts to pack standard Diffusers weights into AWQ/GPTQ formats expected by vLLM kernels.
  - _Research Item_: DiTs often have different activation distributions than LLMs. We need to verify if standard calibration datasets (like Pile) work, or if we need visual calibration datasets (e.g., LAION subsets) for effective quantization.

### DiT-Specific Optimization (P1)
_Goal: Quantize components unique to diffusion._

- Modulation Layer Quantization: DiTs use "Modulation" (adaptive LayerNorm) heavily.
  - Research Item: Can we quantize the ```scale``` and ```shift``` parameters predicted by the timestep embedding?

- TorchAO Integration: Investigate```torchao``` (PyTorch Architecture Optimization) for dynamic quantization of the VAE, which vLLM currently ignores.
  - Target: ```int8``` dynamic quantization for the 3D VAE to reduce memory spikes during video decoding.

### Kernel Research (P2)
_Goal: Custom kernels for visual data._

- SageAttention with INT8: Integrate SageAttention (mentioned in previous roadmaps) with INT8 KV-cache support specifically optimized for 2D/3D attention masks.

- Mixed-Precision DiT: Research "Layer-wise Mixed Precision" where the initial noisy timesteps (which require less precision) use INT4, while the final detailing steps use FP16.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by lishunyang12** [comment]
@SamitHuang @david6666666 @ZJY0516 @hsliuustc0106

**Comment by ZJY0516** [comment]
Thanks for the proposal. It currently serves as a high-level direction or roadmap. To make progress, let's narrow our focus to one concrete quantization target, such as FP8.
From there, we can outline specific, actionable steps — for example, proposing 1-2 viable implementation approaches

**Comment by lishunyang12** [comment]
> Thanks for the proposal. It currently serves as a high-level direction or roadmap. To make progress, let's narrow our focus to one concrete quantization target, such as FP8.
> From there, we can outline specific, actionable steps — for example, proposing 1-2 viable implementation approaches
> 
> 

Thanks for your guideline. I think formulating a concrete plan is important to move it forward. I will draft a separate RFC.

**Comment by BowenBao** [comment]
Thanks for proposal. How much can we reuse the quantization from vLLM? I see that kernel re-use was mentioned but can the entire infrastructure around quantization config, methods, weight/scale loading etc be shared?

**Comment by lishunyang12** [comment]
@BowenBao Good question. The short answer is: a lot is already reused, but there are gaps — especially on the ROCm side.

**What's shared today:**
- Diffusion models that use vLLM's parallel linear layers (`QKVParallelLinear`, `ColumnParallelLinear`, `RowParallelLinear`) get quantization support for free — FP8 already works through this path
- The FP8 config (`vllm_omni/diffusion/quantization/fp8.py`) wraps vLLM's `Fp8Config` directly, including dynamic activation scaling and per-layer skip via `ignored_layers`
- Weight loading uses vLLM's `AutoWeightsLoader` and `default_weight_loader`

**What's missing:**
- ROCm-specific quantization kernel paths are not fully validated — FP8 W8A8 may not fall back correctly on AMD GPUs
- AWQ/GPTQ/INT8 support exists in vLLM but isn't bridged to diffusion models yet
- No testing with quantization + distributed inference (TP, SP, HSDP)

Would you be interested in helping close the ROCm quantization gap? Having someone from the AMD side validate and optimize the FP8 kernel paths for diffusion models on ROCm would be really valuable. Happy to help with context on the diffusion pipeline side if useful.

**Comment by BowenBao** [comment]
hi @lishunyang12 , thanks for sharing the context! We are happy to bring AMD Quark quantization to vllm-omni and enable the FP8 / MXFP4 / etc quantization paths for ROCm. We will work on an RFC and share it here.

**Comment by lishunyang12** [comment]
Superseded by #1854, which is the current quantization roadmap. Original interface design implemented in #1034. Per-component quantization and all alternative methods tracked there.
